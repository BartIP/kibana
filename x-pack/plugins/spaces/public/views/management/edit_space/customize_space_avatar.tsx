/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
// @ts-ignore
import { EuiColorPicker, EuiFieldText, EuiFlexItem, EuiFormRow, EuiLink } from '@elastic/eui';
<<<<<<< HEAD
=======
import { FormattedMessage, InjectedIntl, injectI18n } from '@kbn/i18n/react';
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
import React, { ChangeEvent, Component, Fragment } from 'react';
import { MAX_SPACE_INITIALS } from '../../../../common/constants';
import { Space } from '../../../../common/model/space';
import { getSpaceColor, getSpaceInitials } from '../../../../common/space_attributes';

interface Props {
  space: Partial<Space>;
  onChange: (space: Partial<Space>) => void;
<<<<<<< HEAD
=======
  intl: InjectedIntl;
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
}

interface State {
  expanded: boolean;
  initialsHasFocus: boolean;
  pendingInitials?: string | null;
}

<<<<<<< HEAD
export class CustomizeSpaceAvatar extends Component<Props, State> {
=======
class CustomizeSpaceAvatarUI extends Component<Props, State> {
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
  private initialsRef: HTMLInputElement | null = null;

  constructor(props: Props) {
    super(props);
    this.state = {
      expanded: false,
      initialsHasFocus: false,
    };
  }

  public render() {
    return this.state.expanded ? this.getCustomizeFields() : this.getCustomizeLink();
  }

  public getCustomizeFields = () => {
<<<<<<< HEAD
    const { space } = this.props;
=======
    const { space, intl } = this.props;
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1

    const { initialsHasFocus, pendingInitials } = this.state;

    return (
      <Fragment>
        <EuiFlexItem grow={false}>
<<<<<<< HEAD
          <EuiFormRow label={'Initials (2 max)'}>
=======
          <EuiFormRow
            label={intl.formatMessage({
              id: 'xpack.spaces.management.customizeSpaceAvatar.initialItemsFormRowLabel',
              defaultMessage: 'Initials (2 max)',
            })}
          >
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
            <EuiFieldText
              inputRef={this.initialsInputRef}
              name="spaceInitials"
              // allows input to be cleared or otherwise invalidated while user is editing the initials,
              // without defaulting to the derived initials provided by `getSpaceInitials`
              value={initialsHasFocus ? pendingInitials || '' : getSpaceInitials(space)}
              onChange={this.onInitialsChange}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={true}>
<<<<<<< HEAD
          <EuiFormRow label={'Color'}>
=======
          <EuiFormRow
            label={intl.formatMessage({
              id: 'xpack.spaces.management.customizeSpaceAvatar.colorFormRowLabel',
              defaultMessage: 'Color',
            })}
          >
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
            <EuiColorPicker color={getSpaceColor(space)} onChange={this.onColorChange} />
          </EuiFormRow>
        </EuiFlexItem>
      </Fragment>
    );
  };

  public initialsInputRef = (ref: HTMLInputElement) => {
    if (ref) {
      this.initialsRef = ref;
      this.initialsRef.addEventListener('focus', this.onInitialsFocus);
      this.initialsRef.addEventListener('blur', this.onInitialsBlur);
    } else {
      if (this.initialsRef) {
        this.initialsRef.removeEventListener('focus', this.onInitialsFocus);
        this.initialsRef.removeEventListener('blur', this.onInitialsBlur);
        this.initialsRef = null;
      }
    }
  };

  public onInitialsFocus = () => {
    this.setState({
      initialsHasFocus: true,
      pendingInitials: getSpaceInitials(this.props.space),
    });
  };

  public onInitialsBlur = () => {
    this.setState({
      initialsHasFocus: false,
      pendingInitials: null,
    });
  };

  public getCustomizeLink = () => {
    return (
      <EuiFlexItem grow={false}>
        <EuiFormRow hasEmptyLabelSpace={true}>
          <EuiLink name="customize_space_link" onClick={this.showFields}>
<<<<<<< HEAD
            Customize
=======
            <FormattedMessage
              id="xpack.spaces.management.customizeSpaceAvatar.customizeLinkText"
              defaultMessage="Customize"
            />
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
          </EuiLink>
        </EuiFormRow>
      </EuiFlexItem>
    );
  };

  public showFields = () => {
    this.setState({
      expanded: true,
    });
  };

  public onInitialsChange = (e: ChangeEvent<HTMLInputElement>) => {
    const initials = (e.target.value || '').substring(0, MAX_SPACE_INITIALS);

    this.setState({
      pendingInitials: initials,
    });

    this.props.onChange({
      ...this.props.space,
      initials,
    });
  };

  public onColorChange = (color: string) => {
    this.props.onChange({
      ...this.props.space,
      color,
    });
  };
}
<<<<<<< HEAD
=======

export const CustomizeSpaceAvatar = injectI18n(CustomizeSpaceAvatarUI);
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
