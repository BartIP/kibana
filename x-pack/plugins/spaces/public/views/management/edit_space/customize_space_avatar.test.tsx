/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
// @ts-ignore
import { EuiColorPicker, EuiFieldText, EuiLink } from '@elastic/eui';
<<<<<<< HEAD
import { mount, shallow } from 'enzyme';
import React from 'react';
=======
import React from 'react';
import { mountWithIntl, shallowWithIntl } from 'test_utils/enzyme_helpers';
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
import { CustomizeSpaceAvatar } from './customize_space_avatar';

const space = {
  id: '',
  name: '',
};

test('renders without crashing', () => {
<<<<<<< HEAD
  const wrapper = shallow(<CustomizeSpaceAvatar space={space} onChange={jest.fn()} />);
=======
  const wrapper = shallowWithIntl(
    <CustomizeSpaceAvatar.WrappedComponent space={space} onChange={jest.fn()} intl={null as any} />
  );
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
  expect(wrapper).toMatchSnapshot();
});

test('renders a "customize" link by default', () => {
<<<<<<< HEAD
  const wrapper = mount(<CustomizeSpaceAvatar space={space} onChange={jest.fn()} />);
=======
  const wrapper = mountWithIntl(
    <CustomizeSpaceAvatar.WrappedComponent space={space} onChange={jest.fn()} intl={null as any} />
  );
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
  expect(wrapper.find(EuiLink)).toHaveLength(1);
});

test('shows customization fields when the "customize" link is clicked', () => {
<<<<<<< HEAD
  const wrapper = mount(<CustomizeSpaceAvatar space={space} onChange={jest.fn()} />);
=======
  const wrapper = mountWithIntl(
    <CustomizeSpaceAvatar.WrappedComponent space={space} onChange={jest.fn()} intl={null as any} />
  );
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
  wrapper.find(EuiLink).simulate('click');

  expect(wrapper.find(EuiLink)).toHaveLength(0);
  expect(wrapper.find(EuiFieldText)).toHaveLength(1);
  expect(wrapper.find(EuiColorPicker)).toHaveLength(1);
});

test('invokes onChange callback when avatar is customized', () => {
  const customizedSpace = {
    id: '',
    name: 'Unit Test Space',
    initials: 'SP',
    color: '#ABCDEF',
  };

  const changeHandler = jest.fn();

<<<<<<< HEAD
  const wrapper = mount(<CustomizeSpaceAvatar space={customizedSpace} onChange={changeHandler} />);
=======
  const wrapper = mountWithIntl(
    <CustomizeSpaceAvatar.WrappedComponent
      space={customizedSpace}
      onChange={changeHandler}
      intl={null as any}
    />
  );
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
  wrapper.find(EuiLink).simulate('click');

  wrapper
    .find(EuiFieldText)
    .find('input')
    .simulate('change', { target: { value: 'NV' } });

  expect(changeHandler).toHaveBeenCalledWith({
    ...customizedSpace,
    initials: 'NV',
  });
});
