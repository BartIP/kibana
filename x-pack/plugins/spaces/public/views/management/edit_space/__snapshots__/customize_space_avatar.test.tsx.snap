// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<EuiFlexItem
  component="div"
  grow={false}
>
  <EuiFormRow
    describedByIds={Array []}
    fullWidth={false}
    hasEmptyLabelSpace={true}
  >
    <EuiLink
      color="primary"
      name="customize_space_link"
      onClick={[Function]}
      type="button"
    >
<<<<<<< HEAD
      Customize
=======
      <FormattedMessage
        defaultMessage="Customize"
        id="xpack.spaces.management.customizeSpaceAvatar.customizeLinkText"
        values={Object {}}
      />
>>>>>>> ff49a1c6742d67fa5daed569ff3bb269783f6bd1
    </EuiLink>
  </EuiFormRow>
</EuiFlexItem>
`;
